name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/util/chart/ChartToolsManager.java

```diff
@@ -0,0 +1,523 @@
+
+            Point rectPoint = new Point(e.width - txtYExtend.x - 2, 0);
+            Point textPoint = new Point(e.width - txtYExtend.x + 1, 0);
+
+            // Visual shift horizontally of the container
+            if (p.y <= e.height / 2)
+            {
+                rectPoint.y = p.y + 4;
+                textPoint.y = p.y + 7;
+            }
+            else
+            {
+                rectPoint.y = p.y - txtYExtend.y - 4;
+                textPoint.y = p.y - txtYExtend.y - 1;
+            }
+
+            e.gc.setBackground(Colors.brighter(color));
+            e.gc.setForeground(Colors.getTextColor(color));
+
+            e.gc.fillRoundRectangle(rectPoint.x, rectPoint.y, txtYExtend.x, txtYExtend.y, PADDING, PADDING);
+            e.gc.drawText(yText, textPoint.x, textPoint.y, true);
+        }
+
+        private void drawCrosshairValueSecondAxisTextbox(PaintEvent e, Point p)
+        {
+            IAxis axis = chart.getAxisSet().getYAxis(2);
+            if (!axis.getTick().isVisible())
+                return;
+
+            double value = axis.getDataCoordinate(p.y);
+
+            String yText = Values.PercentWithSign.format(value);
+
+            // Add margin to text
+            Point txtYExtend = e.gc.textExtent(yText);
+            txtYExtend.x += 5;
+            txtYExtend.y += 5;
+
+            Point rectPoint = new Point(1, 0);
+            Point textPoint = new Point(3, 0);
+
+            // Visual shift horizontally of the container
+            if (p.y <= e.height / 2)
+            {
+                rectPoint.y = p.y + 4;
+                textPoint.y = p.y + 7;
+            }
+            else
+            {
+                rectPoint.y = p.y - txtYExtend.y - 4;
+                textPoint.y = p.y - txtYExtend.y - 1;
+            }
+
+            e.gc.setBackground(Colors.brighter(color));
+            e.gc.setForeground(Colors.getTextColor(color));
+
+            e.gc.fillRoundRectangle(rectPoint.x, rectPoint.y, txtYExtend.x, txtYExtend.y, PADDING, PADDING);
+            e.gc.drawText(yText, textPoint.x, textPoint.y, true);
+        }
+
+        private void drawCrosshair(PaintEvent e, Point p)
+        {
+            e.gc.setLineWidth(1);
+            e.gc.setLineStyle(SWT.LINE_SOLID);
+            e.gc.setForeground(color);
+            e.gc.setBackground(color);
+            e.gc.setAntialias(SWT.ON);
+
+            // draw crosshair
+            e.gc.drawLine(p.x, 0, p.x, e.height);
+            e.gc.drawLine(0, p.y, e.width, p.y);
+
+            // set textbox style
+            e.gc.setForeground(Colors.theme().defaultForeground());
+            e.gc.setBackground(Colors.theme().defaultBackground());
+            e.gc.setAlpha(220);
+
+            // value for horizontal axis
+            drawCrosshairDateTextbox(e, p);
+
+            drawCrosshairValueTextbox(e, p);
+            drawCrosshairValueSecondAxisTextbox(e, p);
+        }
+    }
+
+    private static class MeasurementTool extends ChartToolBase
```

Comment:
You moved the MeasurementTool from a separate file into a member class.
I also tend to separate classes into smaller pieces rather later (there are some big classes 🫣), but I think the measurement and cross hair tools can have their own files - for readability and maintenance reasons.


If you feel like building more tools, we could also add a "tools" packages below the "chart" package.

Reply:
Totally agree.